draft-ietf-quic-tls.md

```diff
@@ -613,16 +613,27 @@ by the "initial_max_data" transport parameter supplied by the server.  A client
 MUST treat receipt of a NewSessionTicket that contains an "early_data" extension
 with any other value as a connection error of type PROTOCOL_VIOLATION.
 
+A client that wishes to send early data uses the "early_data" extension in the
+ClientHello message of a subsequent handshake (see Section 4.10.2 of
+{{!TLS13}}). It then sends the application data in 0-RTT packets.
+
 Early data within the TLS connection MUST NOT be used.  As it is for other TLS
 application data, a server MUST treat receiving early data on the TLS connection
 as a connection error of type PROTOCOL_VIOLATION.
 
 
-## Rejecting 0-RTT
+## Accepting and rejecting 0-RTT
+
+A server accepts 0-RTT by sending the EarlyDataIndication in the ServerHello
+(see Section 4.10.2 of {{!TLS13}}).  The server then processes and acknowledges
+the 0-RTT packets that it receives.
 
-A server rejects 0-RTT by rejecting 0-RTT at the TLS layer.  This also prevents
-QUIC from sending 0-RTT data. A server will always reject 0-RTT if it sends a
-TLS HelloRetryRequest.
+A server rejects 0-RTT by sending a ServerHello without the EarlyDataIndication.
+A server will always reject 0-RTT if it sends a TLS HelloRetryRequest.  When
+rejecting 0-RTT, a server MUST NOT process any 0-RTT packets, even if it is in
+possesion of the keys to do so.  When 0-RTT was rejected, a client MUST treat
+receipt of an acknowledgement for a 0-RTT packet as a connection error of type
```

Comment:
I wonder if this needs to be a MUST (in sense that the client is required to track the packet numbers it has used for 0-RTT packets, even when it knows that the server has rejected 0-RTT).

Reply:
I don't think you need to explicitly track the packet numbers. Here's how I would implement this:
When 0-RTT is rejected, I clear all recovery state for the 0-RTT packets that I sent earlier. Now when receiving an ACK, I just need to run the standard check that it only acknowledges packets that I actually sent. No special case is required to handle this MUST.

Reply:
To clarify, what is being discussed here is about how to process ACKs containing invalid packet numbers. The general rule we have is that _an endpoint SHOULD treat receipt of an acknowledgment for a packet it did not send as a connection error of type PROTOCOL_VIOLATION, if it is able to detect the condition_ ([section 13.1.1](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#rfc.section.13.1.1)).

Is there a reason we need something different for 0-RTT packets?

Reply:
Makes sense. I updated the text.

Reply:
Can we have "if it is able to detect the condition"?

I do not think we need to recommend all clients implement the check only for this particular purpose.

Reply:
I'm OK with this MUST.  There are ways to manage this that work reasonably well.

Reply:
@martinthomson I disagree that is is easy to detect.

It is true that it is easy for a QUIC stack that uses different packet number spaces to detect this error. This is because all you need to do is remember the largest PN used for the 0-RTT packet, and compare that with ACKs you receive in the application space.

OTOH, this is difficult for a QUIC stack that uses unified packet number space. Such stacks would be required to remember all the PNs being used for sending 0-RTT packets.

The other question is if we need to require endpoints to posses the 0-RTT keys, even after it detects that 0-RTT has been rejected. Unless we require that, having this MUST becomes much useless.

Speaking of text, I think there are four options:
a) MUST detect and close
b) SHOULD detect and close
c) MUST close if detected
d) SHOULD close if detected

As stated, I oppose to _a_ because it could be hard to detect. I think _c_ is fine. I might agree that _b_, _d_ might not be good, though I wonder what's the necessity to have something different from the general rule of handling ACKs carrying invalid PNs. Maybe the general rule needs to be changed to MUST.

Reply:
@kazuho There's no such thing as a unified PN space. QUIC has 3 packet number spaces, and there's no way to opt out of that. If a QUIC implementation decides to do weird things with the packet numbers (i.e. if it doesn't start every PN space from 0, as the spec suggests), it shouldn't be surprised to run into additional complexity...

That being said, as you point out in https://github.com/quicwg/base-drafts/pull/2841#discussion_r297128544, we're using a SHOULD elsewhere, so for consistency I changed it to a SHOULD here as well. I don't think that we need a "if it is able to detect the condition" sub-clause here, as the SHOULD already allows you to skip the check, if your implementation is not able to perform it.

Reply:
@marten-seemann The spec having different packet number spaces does not necessarily mean that every QUIC stack should be having different PN counters for each number space for the packets it sends. It is entirely plausible for a QUIC stack to use just one packet number space for the packets it sends.

That said, ...

> as you point out in https://github.com/quicwg/base-drafts/pull/2841#discussion_r297128544, we're using a SHOULD elsewhere, so for consistency I changed it to a SHOULD here as well.

Thank you for the change. I appreciate it.

> I don't think that we need a "if it is able to detect the condition" sub-clause here, as the SHOULD already allows you to skip the check, if your implementation is not able to perform it.

IIRC, the sub-clause "if it is able to detect the condition" was added intentionally to the general rule, because people were concerned about the text being interpreted as to recommend having a way to detect such errors. I think the text proposed in this PR has the same problem, and my preference is to address it in the same manner.

I think that one path forward might be to, instead of clarifying the action here, incorporate the handling of acknowledgements of 0-RTT packets into the provision of 13.1.1.

I mean, changing
> An endpoint SHOULD treat receipt of an acknowledgment for a packet it did not send as a connection error of type PROTOCOL_VIOLATION, if it is able to detect the condition.

to something like:
> An endpoint SHOULD treat receipt of an acknowledgment for a packet that the server would not have processed as a connection error of type PROTOCOL_VIOLATION, if it is able to detect the condition. This includes receiving an ACK frame containing a packet number that the endpoint has not sent, as well as acknowledgements for 0-RTT packets when the server has rejected the use of 0-RTT.

Reply:
@kazuho
> The spec having different packet number spaces does not necessarily mean that every QUIC stack should be having different PN counters for each number space for the packets it sends. It is entirely plausible for a QUIC stack to use just one packet number space for the packets it sends.

The spec is actually quite clear about this (https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#rfc.section.12.3)
>  Packet numbers in each space start at packet number 0.

Now I agree that it's possible to ignore this sentence and start the packet number wherever you want (if you're willing to tolerate the loss recovery response this provokes, i.e. an early ACK because of detected reordering). However, I don't think we need to modify the spec because of self-inflicted complexities of non-compliant implementations.

> IIRC, the sub-clause "if it is able to detect the condition" was added intentionally to the general rule, because people were concerned about the text being interpreted as to recommend having a way to detect such errors.

I think we added this text because for 1-RTT packets, you clear out old ACK ranges from time to time. When receiving an ACK for a very old packet you'd then not be able to check if you actually sent this packet, because you already removed the tracking state.

Reply:
@marten-seemann Contrary to what you state, the spec is clear about skipping PN 0 (or any packet number in a packet number space) is a permitted behavior. Quoting from [section 21.3](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#rfc.section.21.3), _an endpoint MAY skip packet numbers when sending packets to detect this behavior._

Using unified packet number space on the send side is precisely an act of "skipping packet numbers when sending packets," skipping those that have been used in another packet number space.

>> IIRC, the sub-clause "if it is able to detect the condition" was added intentionally to the general rule, because people were concerned about the text being interpreted as to recommend having a way to detect such errors.
> I think we added this text because for 1-RTT packets, you clear out old ACK ranges from time to time. When receiving an ACK for a very old packet you'd then not be able to check if you actually sent this packet, because you already removed the tracking state.

I'm not sure if I agree with the assumption, but granted that it is true, I think it would be an argument _for_ having the same sub-clause for 0-RTT packets as well. ACK ranges are cleared out from time to time. Then, why do we need to handle ranges for 0-RTT packets specially, only when the server rejects 0-RTT?

Reply:
@kazuho Of course you **can** skip packet numbers, since this is indistinguishable from packet loss.
The text you're quoting is about the optimistic ACK attack, and describes a mechanism how you skip a single packet number from time to time **in order to detect this attack**. It is not intended to start a packet number space at a much higher packet number. Furthermore, if I understand your point here correctly, what you actually want to do is **not** to perform the check if a skipped packet number is acknowledged, so you really can't use this text as a justification...

Reply:
Using unified packet number space on the send side is an easy way of detecting an optimistic ACK attack. Because all you need is a single ACK map (instead of ACK map per every packen number space), that allows each of the entry to have the epochs associated. Optimistic ACK is detected when you receive an ACK in an unexpected epoch.

> Furthermore, if I understand your point here correctly, what you actually want to do is not to perform the check if a skipped packet number is acknowledged, so you really can't use this text as a justification...

I think you are missing the fact that there are two topics being discussed in parallel. One is if use of unified packet number space is OK. That relates to if the detection could be MUST. For that, we are in agreement that it can be SHOULD. Therefore, I think it's getting off-topic.

The other discussion is if we need the "if detected" sub-clause. I do not think I am using the existence of QUIC stacks using unified packet number space on the send side as a justification to argue for having the sub-clause.

Reply:
I think the suggestion you made in https://github.com/quicwg/base-drafts/pull/2841#discussion_r297452432 is a good one. I updated the PR.